fix: improper syntax that will be rejected post-2.1 (#1417)

mosteo opened the pull request:
We had a missing check in index loading logic that should have rejected an unexpected array.

Fabien-Chouteau reviewed (commented):
I see why you want to do this, we regularly have questions from people that never update the index. However I am a but worried about the consequences :)

What happens if there's no internet connection?

mosteo commented:
Actually I thought that you also wanted it :) given the discussion in #1348.

Yup, I forgot to test what happens in the offline case. We should fail the update gracefully with a warning and continue as before. We probably don't want to try again immediately but back-off for the configured period in any case. I'll do a bit more testing.

rowan-walshe commented:
I like the change. I was once on of those that had to ask why alr couldn't find a newer version of a crate that I could see in the index on GitHub.

One comment I have is that while `alr update --online` can be used to force update the index(es) before attempting crate updates, it feels like it should still call `Auto_Update_Index` if `--online` isn't set

mosteo commented:
Updated to ensure that an automatic update failure causes the regular backoff, and is handled as a warning.

Also added a test for it. I didn't find a simple solution to make the test locally, so that test uses an [online repo](https://github.com/alire-project/test-index) I created to that effect. There may be solutions around that by using a local git server, but that seemed too complex just to avoid going online, given that that's not a crucial situation for our testing.
